### PR TITLE
openconnect: add support for pkcs11 smartcards

### PR DESCRIPTION
**Pull request by trobert**
###### Motivation for this change
Allow to use pkcs11 smartcards with openconnect.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [X] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @pradeepchhetri

### CODE REVIEW

**Review by tricktron — APPROVED** [review]
Looks good. I could successfully build this on linux.

**Comment by trobert** [comment]
@FRidh / @LnL7 Does this need more reviewing or can it merged ?

**Comment by LnL7** [comment]
Looks good to me, but similar to #71318 I don't think any of these dependencies have to be propagated. Nix will figure out what are runtime dependencies automatically.

**Comment by trobert** [comment]
I think the reason why the dependencies are propagated is because of the libopenconnect shared library. Is there a recipe to figure out which dep to propagate ?

**Comment by LnL7** [comment]
The only case where this could help is if _every_ downstream project that links against the package will also have to link against the dependency directly. An example of where we do need propagated inputs are python packages, unlike shared libraries all the dependencies need to be available in PYTHONPATH there, not just direct dependencies of the package.

Here's an example of what nix figures out automatically, based on what binaries and libraries link against during the build.

```
$ nix why-depends -f . openconnect openssl.out
/nix/store/siqwim1myqy03mr7h9yi75yb1vnj7mi8-openconnect-8.05
╚═══bin/openconnect: ….h.............8...8./nix/store/lvcx6lyi99lb8xz91qzviy7m499hq1vn-gnutls-3.6.9/lib/libg
    => /nix/store/lvcx6lyi99lb8xz91qzviy7m499hq1vn-gnutls-3.6.9
    ╚═══lib/libgnutls-dane.0.dylib: ….p.................../nix/store/9yynz4296xn6xc1jqxc6z95f3h2nr7cw-unboun
        => /nix/store/9yynz4296xn6xc1jqxc6z95f3h2nr7cw-unbound-1.9.3-lib
        ╚═══lib/libunbound.8.dylib: ….h.................../nix/store/iwimdjfsf037k9abgnvqpcafv893wf2n-openss
            => /nix/store/iwimdjfsf037k9abgnvqpcafv893wf2n-openssl-1.1.1d
```

**Comment by trobert** [comment]
I think nobody links against libopenconnect anyway(only the binay is used).
so I'll move everything to buildInputs.

**Comment by FRidh** [comment]
Please try moving deps to `buildInputs`, though do it in a second commit on this PR. If we end up with issues we can revert that second commit and still have the requested pkcs11 support.

**Comment by trobert** [comment]
deps were already moved to `buildInputs` in 43a761da, so I just rebased my commit on last master.

Is it still in time for 20.03 ?

**Review by tricktron — APPROVED** [review]
Looks fine.

**Comment by trobert** [comment]
@FRidh: Do you think this can this be merged ? Thanks

**Comment by SuperSandro2000** [comment]
@trobert can you please target staging with this PR? It is causing quite some rebuilds.

**Review by tricktron — APPROVED** [review]
Looks good to me

**Comment by tricktron** [comment]
@SuperSandro2000 Can this be merged?

File: pkgs/tools/networking/openconnect/default.nix [hunk 1]
```diff
@@ -1,6 +1,7 @@
-{ lib, stdenv, fetchurl, pkg-config, openssl ? null, gnutls ? null, gmp, libxml2, stoken, zlib, fetchgit, darwin } :
+{ lib, stdenv, fetchurl, pkgconfig, openssl ? null, gnutls ? null, gmp, libxml2, p11-kit, libp11 ? null, stoken, zlib, fetchgit, darwin } :
```
Comment (SuperSandro2000) [inline pkgs/tools/networking/openconnect/default.nix:1]:
```suggestion
{ lib, stdenv, fetchurl, pkg-config, openssl ? null, gnutls ? null, gmp, libxml2, p11-kit, libp11, stoken, zlib, fetchgit, darwin } :
```

File: pkgs/tools/networking/openconnect/default.nix [hunk 2]
```diff
@@ -27,9 +28,10 @@ in stdenv.mkDerivation rec {
     "--without-openssl-version-check"
   ];
 
-  buildInputs = [ openssl gnutls gmp libxml2 stoken zlib ]
-    ++ lib.optional stdenv.isDarwin darwin.apple_sdk.frameworks.PCSC;
-  nativeBuildInputs = [ pkg-config ];
+  buildInputs = [ openssl gnutls gmp libxml2 p11-kit stoken zlib ]
+    ++ lib.optional stdenv.isDarwin darwin.apple_sdk.frameworks.PCSC
+    ++ lib.optional (openssl != null) libp11;
+  nativeBuildInputs = [ pkgconfig ];
```
Comment (SuperSandro2000) [inline pkgs/tools/networking/openconnect/default.nix:34]:
```suggestion
  nativeBuildInputs = [ pkg-config ];
```

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]

File: pkgs/tools/networking/openconnect/default.nix [hunk 3]
```diff
@@ -27,8 +28,9 @@ in stdenv.mkDerivation rec {
     "--without-openssl-version-check"
   ];
 
-  buildInputs = [ openssl gnutls gmp libxml2 stoken zlib ]
-    ++ lib.optional stdenv.isDarwin darwin.apple_sdk.frameworks.PCSC;
+  buildInputs = [ openssl gnutls gmp libxml2 p11-kit stoken zlib ]
+    ++ lib.optional stdenv.isDarwin darwin.apple_sdk.frameworks.PCSC
```
Comment (SuperSandro2000) [inline pkgs/tools/networking/openconnect/default.nix:32]: darwin shouldn't be used as an input.

Comment (trobert) [reply pkgs/tools/networking/openconnect/default.nix:32]: you're saying that we should remove the darwin dependency ?
Note that I don't have the possibility to test under darwin...

Comment (SuperSandro2000) [reply pkgs/tools/networking/openconnect/default.nix:32]: I mean like this https://github.com/NixOS/nixpkgs/blob/7248f89da2dab8bca3a9f4f11d0407d0c8e25e89/pkgs/top-level/all-packages.nix#L2899

File: pkgs/tools/networking/openconnect/default.nix [hunk 4]
```diff
@@ -1,6 +1,7 @@
-{ lib, stdenv, fetchurl, pkg-config, openssl ? null, gnutls ? null, gmp, libxml2, stoken, zlib, fetchgit, darwin } :
+{ lib, stdenv, fetchurl, pkg-config, openssl ? null, gnutls ? null, gmp, libxml2, p11-kit, libp11, stoken, zlib, fetchgit, darwin } :
```
Comment (tricktron) [inline pkgs/tools/networking/openconnect/default.nix:1]: Replace `darwin` here with `PCSC`

**Review by tricktron — CHANGES_REQUESTED** [review]

File: pkgs/tools/networking/openconnect/default.nix (hunk 4 above)
Comment (tricktron) [reply pkgs/tools/networking/openconnect/default.nix:1]: @roberth Could you add these changes?
@SuperSandro2000 I guess it makes more sense to first merge this pr before refactoring the `vpnc-script` into its own module as suggested https://github.com/NixOS/nixpkgs/pull/106465.

Comment (roberth) [reply pkgs/tools/networking/openconnect/default.nix:1]: I'm not using/maintaining smart card packages anymore. I think you meant @trobert :)

Comment (tricktron) [reply pkgs/tools/networking/openconnect/default.nix:1]: @roberth Sorry about that and thanks for letting me know. I edited my comment.
At least I was close with the name confusion😃.

File: pkgs/tools/networking/openconnect/default.nix [hunk 5]
```diff
@@ -1,6 +1,7 @@
...
 
 assert (openssl != null) == (gnutls == null);
+assert (openssl != null) -> (libp11 != null);
```
Comment (SuperSandro2000) [inline pkgs/tools/networking/openconnect/default.nix:4]:
```suggestion
```

we don't need this. libp11 cannot be null.

File: pkgs/tools/networking/openconnect/default.nix (hunk 4 above)
Comment (SuperSandro2000) [reply pkgs/tools/networking/openconnect/default.nix:1]: Yeah, sure. We can do that.

File: pkgs/tools/networking/openconnect/default.nix (hunk 5 above)
Comment (trobert) [reply pkgs/tools/networking/openconnect/default.nix:4]: I've now made it null when build with openssl

**Comment by superherointj** [comment]
@trobert Please fix the conflict. And report current situation.

File: pkgs/tools/networking/openconnect/default.nix (hunk 5 above)
Comment (SuperSandro2000) [reply pkgs/tools/networking/openconnect/default.nix:4]: We should use boolean inputs instead of overwriting packages to null

**Comment by SuperSandro2000** [comment]
Closing because last author activity was 15 months ago.

**Pull request closed without merging**